File: pythonworkspace/quick_sort.py
arr=[12,4,2,9,23,22,15,0,10]

def partition(arr, i,j):

    p_index=j
    left=i 
    right=j-1
    while left<right:
        while arr[left]<arr[p_index]:
            left+=1
        while arr[right]>arr[p_index]:
            right-=1
        if right<=left:
            break
        arr[left],arr[right]=arr[right],arr[left]
        left+=1
        right-=1
    final_piv=right+1   
    arr[final_piv],arr[p_index]=arr[p_index],arr[final_piv]
    return final_piv

def quicksort(arr,i,j):
    if i<j:
        final_piv=partition(arr, i, j)
        quicksort(arr, i, final_piv-1)
        quicksort(arr, final_piv+1,j)
    

quicksort(arr, 0, 8)
print(arr)

File: Quicksort.java
import java.util.Scanner;
import java.io.*;

public class Quicksort{
	static int counter=0;

	static void swap(int[] arr, int left, int right){
		int temp=arr[left];
		arr[left]=arr[right];
		arr[right]=temp;
		if(arr[left]<arr[right])
		++counter;
	}

	public static void qsort(int[] arr, int left, int right, int N){
		if(left>=right) return;

		int part=qpart(arr,left,right,N);
		qsort(arr,left,part-1,N);
		qsort(arr,part+1,right,N);
	}

	public static int qpart(int[] arr, int left, int right, int N){
	
		int highstart=left-1,pivot=right;
		int pivotval=arr[pivot];

		for(int iter=left;iter<right;iter++){
			if(arr[iter]<=pivotval) {
				swap(arr,++highstart,iter);
			}
		}

		swap(arr,++highstart,pivot);

		return highstart;

	}

	public static void main(String[] args) {
		Scanner box= new Scanner(System.in);
		 final int N=box.nextInt();
		int[] input= new int[N];

		for (int i=0; i<N; i++) {
			input[i]=box.nextInt();
		}

		qsort(input,0,N-1,N);

//		int a = qpart(input,0,N-1);

		for (int i=0; i<N; i++) {
			System.out.print(input[i]+" ");
		}

		System.out.println("\npossible counter  :"+ counter);



	}
}